includes/AMP/Traits/Sanitization_Utils.php

```diff
@@ -113,6 +115,17 @@ private function transform_a_tags( $document ): void {
 			if ( ! $link->getAttribute( 'data-tooltip-text' ) ) {
 				$link->removeAttribute( 'data-tooltip-text' );
 			}
+
+			// Extra hardening to catch links without a proper protocol.
+			// Matches withProtocol() util in the editor.
+			if (
+				0 !== stripos( $url, 'http://' ) &&
+				0 !== stripos( $url, 'https://' ) &&
+				0 !== stripos( $url, 'tel:' ) &&
+				0 !== stripos( $url, 'mailto:' )
```

Comment:
Could we use `wp_allowed_protocols()` here?

Reply:
Maybe look at `wp_kses_bad_protocol_once`.

Reply:
This matches the list of allowed protocols in the editor, as per the inline comment.

Reply:
Okay, how about an array you loop around then?

Reply:
Again we can use `str_starts_with` here.

Reply:
You mean something like this?

```php
$has_protocol = false;
foreach ( [ 'http://', 'https://', 'tel:', 'mailto:' ] as $protocol ) {
	if ( 0 === stripos( $url, $protocol ) ) {
		$has_protocol = true;
		break;
	}
}

if ( ! $has_protocol ) {
	$link->setAttribute( 'href', 'https://' . $url );
}
```

That is much longer and less readable

Reply:
I find the more readable. But we have had this discussion before..